### PR TITLE
Add fg-loadcss type definitions

### PR DESCRIPTION
**Pull request by ncovercash**
Original repo: https://github.com/filamentgroup/loadCSS

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/fg-loadcss/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// Type definitions for fg-loadcss 3.1
+// Project: https://github.com/filamentgroup/loadCSS
+// Definitions by: Noah Overcash <https://github.com/ncovercash>
```
Comment (danger-public) [inline types/fg-loadcss/index.d.ts:3]: - :no_entry_sign: This line should have your github username in it, not /me

**Comment by danger-public** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **types/fg-loadcss/fg-loadcss-tests.ts#L9** - The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 316519472409a955f1bd5fd02f41835a65dbda60
</p>

File: types/fg-loadcss/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export function loadCSS(href: string, before?: HTMLElement, media?: string, attributes?: Record<string, string>): HTMLLinkElement;
+
+export function onloadCSS(stylesheet: HTMLLinkElement, callback: () => void): void;
```
Comment (peterblazejewicz) [inline types/fg-loadcss/index.d.ts:8]: this one should be probably defined on window:
```ts
declare namespace global {
   interface Window {
        onloadCSS(....
````
it's optional, not part of exported UMD:
https://github.com/filamentgroup/loadCSS/blob/master/src/onloadCSS.js#L2
https://unpkg.com/fg-loadcss@3.1.0/src/loadCSS.js

File: types/fg-loadcss/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// Type definitions for fg-loadcss 3.1
+// Project: https://github.com/filamentgroup/loadCSS
+// Definitions by: Noah Overcash <https://github.com/ncovercash>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export function loadCSS(href: string, before?: HTMLElement, media?: string, attributes?: Record<string, string>): HTMLLinkElement;
```
Comment (peterblazejewicz) [inline types/fg-loadcss/index.d.ts:6]: AFAIK, this seems to be UMD module, so this one would require minor change:
https://unpkg.com/fg-loadcss@3.1.0/src/loadCSS.js
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#identifying-a-umd-library
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#examples-of-umd-libraries

```ts
export as namespace loadCSS ...

declare function loadCSS...

export = loadCSS;
````

File: types/fg-loadcss/fg-loadcss-tests.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { loadCSS, onloadCSS } from "fg-loadcss";
```
Comment (peterblazejewicz) [inline types/fg-loadcss/fg-loadcss-tests.ts:1]: It cannot work that way, the 'onloadCSS' IMO is not a part of a module, can you dlb check this?

File: types/fg-loadcss/index.d.ts (hunk 2 above)
Comment (ncovercash) [reply types/fg-loadcss/index.d.ts:8]: Gotcha, thanks!

**Comment by ncovercash** [comment]
@peterblazejewicz I believe I fixed the issues you mentioned.  To be able to keep the namespaces separate I tried this which, per my tests, works, however, I'm not certain of what I'm doing.  I'd really appreciate another review!

**Comment by ncovercash** [comment]
> I believe I fixed the issues you mentioned. To be able to keep the namespaces separate I tried this which, per my tests, works, however, I'm not certain of what I'm doing. I'd really appreciate another review!

Ok never mind, I have no idea what I'm doing – any suggestions?

File: types/fg-loadcss/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+declare namespace FgLoadcss {
+    function loadCSS(href: string, before?: HTMLElement, media?: string, attributes?: Record<string, string>): HTMLLinkElement;
+}
+
+declare module "fg-loadcss" {
+    export = FgLoadcss;
+}
+
+declare var onloadCSS: (stylesheet: HTMLLinkElement, callback: () => void) => void;
```
Comment (peterblazejewicz) [inline types/fg-loadcss/index.d.ts:14]: try this:
```suggestion
export as namespace loadCSS;

declare function loadCSS(
    href: string,
    before?: HTMLElement,
    media?: string,
    attributes?: Record<string, string>,
): HTMLLinkElement;

export = loadCSS;

declare global {
    interface Window {
        onloadCSS: (stylesheet: HTMLLinkElement, callback: () => void) => void;
    }
}
```

File: types/fg-loadcss/fg-loadcss-tests.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { loadCSS } from "fg-loadcss";
```
Comment (peterblazejewicz) [inline types/fg-loadcss/fg-loadcss-tests.ts:1]:
```suggestion
import loadCSS  = require("fg-loadcss");
```
for CommonJS usage

File: types/fg-loadcss/fg-loadcss-tests.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+loadCSS("example.css");
+loadCSS("example.css", document.body);
+loadCSS("example.css", document.body, "print");
+const foo = loadCSS("example.css", document.body, "print", { integrity: "sha123-abc" });
+
+onloadCSS(foo, () => {
```
Comment (peterblazejewicz) [inline types/fg-loadcss/fg-loadcss-tests.ts:8]:
```suggestion
window.onloadCSS(foo, () => {
```

**Comment by ncovercash** [comment]
Thank you so much!

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## fg-loadcss ([<kbd>unpkg</kbd>](https://unpkg.com/browse/fg-loadcss@latest/))
was missing the following properties:
1. loadCSS
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 532db36ae1341d7a5014ee5986b34a91318fe121
</p>

**Comment by ncovercash** [comment]
@peterblazejewicz from my usage, I have had to use `import { loadCSS } from "fg-loadcss";` and it looks like there is no default export.  How should I reflect that with the namespace style?

**Comment by peterblazejewicz** [comment]
let me dlb check

File: types/fg-loadcss/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+export as namespace loadCSS;
+
+declare function loadCSS(
+    href: string,
+    before?: HTMLElement,
+    media?: string,
+    attributes?: Record<string, string>,
+): HTMLLinkElement;
+
+export = loadCSS;
```
Comment (peterblazejewicz) [inline types/fg-loadcss/index.d.ts:15]:
```suggestion
export function loadCSS(
    href: string,
    before?: HTMLElement,
    media?: string,
    attributes?: Record<string, string>,
): HTMLLinkElement;
```
you're correct, this works. There is only single, named export in this module

File: types/fg-loadcss/fg-loadcss-tests.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import loadCSS  = require("fg-loadcss");
```
Comment (peterblazejewicz) [inline types/fg-loadcss/fg-loadcss-tests.ts:1]:
```suggestion
import { loadCSS } from 'fg-loadcss';
```

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@ncovercash sorry for the fuss, I've got this wrong from a module,
still quite sure about that additional function, it's just cannot be part of the
module itself (comes as an extra)

**Pull request merged**